File: android/src/main/kotlin/com/retozu/flutter_amazon_chime/MetricsObserver.kt
/*
 * Copyright (c) 2025 Duong Le. MIT License.
 */

package com.retozu.flutter_amazon_chime

import com.amazonaws.services.chime.sdk.meetings.audiovideo.metric.MetricsObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.metric.ObservableMetric

class ChimeMetricsObserver(private val chimeFlutterApi: ChimeFlutterApi) : MetricsObserver {
    override fun onMetricsReceived(metrics: Map<ObservableMetric, Any>) {
        val stringMetrics: Map<String?, Any?> = metrics.entries.associate { (key, value) -> key.name to value }
        chimeFlutterApi.onMeetingMetricsReceived(stringMetrics) {}
    }
}

File: ios/Classes/MetricsObserver.swift
import AmazonChimeSDK

class ChimeMetricsObserver: MetricsObserver {
    let chimeFlutterApi: ChimeFlutterApi

    init(chimeFlutterApi: ChimeFlutterApi) {
        self.chimeFlutterApi = chimeFlutterApi
    }

    func metricsDidReceive(metrics: [AnyHashable: Any]) {
        var stringMetrics = [String: Any?]()
        for (key, value) in metrics {
            stringMetrics[String(describing: key)] = toPigeonValue(value)
        }
        chimeFlutterApi.onMeetingMetricsReceived(metrics: stringMetrics) { _ in }
    }

    private func toPigeonValue(_ value: Any) -> Any? {
        switch value {
        case let s as String: return s
        case let n as NSNumber: return n
        case let b as Bool: return b
        default: return String(describing: value)
        }
    }
}
